### ISSUE TITLE
potential issues with conditions for output writing

### ISSUE THREAD

**Issue by echoi** [tan2/DynEarthSol#2]
I think I've found two potential issues in conditions for output writing. I say "potential" because I noticed them only in the debug mode.
The following codes are from lines 365-369 in `dynearthsol.cxx`:
```C++
if ((! param.sim.output_averaged_fields || (var.steps % param.sim.output_averaged_fields == 0)) &&
            // When output_averaged_fields in turned on, the output cannot be
            // done at arbitrary time steps.
            (((var.steps - starting_step) == next_regular_frame * param.sim.output_step_interval) ||
             ((var.time - starting_time) > next_regular_frame * param.sim.output_time_interval_in_yr * YEAR2SEC)) ) {
```

1. In line 365, `var.steps % param.sim.output_averaged_fields` is not defined (or implementation-dependent) when `output_averaged_fields` is `0`. `output_averaged_fields` can retain the switching function (boolean or 0/1) but a new parameter will have to be used for an averaging interval.
2. In line 368, `param.sim.output_step_interval` is set to be the system's max integer when not specified in an input file. Then, when `next_regular_frame` is greater than 1, this line can cause an overflow. 
      - Line 369 has the same risk because `param.sim.output_time_interval_in_yr` is set to be the system's maximum double-precision number when not specified in an input file.

**Comment by tan2** [comment]
Your analysis on `param.sim.output_averaged_fields` is correct. 

I don't use `param.sim.output_averaged_fields` for anything other than 0 or 1. Do you know anyone use it that way?  

I plan to change `param.sim.output_averaged_fields` to a `bool` and rename it to `param.sim.is_outputting_averaged_fields` for the sake of consistency.


The 2nd issue can be fixed by checking `param.sim.output_step_interval == 
 std::numeric_limits<int>::max()` additionally. Will fix it soon.

**Comment by echoi** [comment]
I haven't used the averaging option extensively either but lines 570-579 of input.cxx. seem to imply that `param.sim.output_averaged_fields` can be greater than 1.
Your planned fixes seem proper. I'll look forward to them.

**Comment by tan2** [comment]
The first problem is fixed in 66a0ede2ec33.
The second problem is fixed in 0dcf5294251e.

**Comment by tan2** [comment]
The first fix has a bug, which results in corrupted (averaged) stress and velocity output.
It is fixed in 4844da3ced8e7b4e7b7c960dae7a7bf8d7556c16.
